# KPI 3 - Driver Lap Times - Data Validation and Sanity Checks

In [12]:
import os 
current = os.getcwd()
print(current)

/Users/frankdong/Documents/Analytics Local/williams-racing-strategies/notebooks


In [ ]:
import pandas as pd

# read csv file
df_laptimes = pd.read_csv('../processed_data/driver-lap-times.csv')

# dataframe basic info
print(df_laptimes.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8222 entries, 0 to 8221
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   race_id                8222 non-null   int64 
 1   gp_year                8222 non-null   int64 
 2   gp_name                8222 non-null   object
 3   gp_round               8222 non-null   int64 
 4   driver_id              8222 non-null   int64 
 5   driver_name            8222 non-null   object
 6   rookie_or_experienced  8222 non-null   object
 7   lap_number             8222 non-null   int64 
 8   lap_time               8222 non-null   object
 9   lap_time_ms            8222 non-null   int64 
dtypes: int64(6), object(4)
memory usage: 642.5+ KB
None


## Summary of processed dataset 'driver-lap-times.csv'

- Filepath: /Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/driver-lap-times.csv *(potentially fix from absolute to relative path later?)*
- Range: 8222 entries, 0 to 8221.
- Columns: 10
- Data types: int64(6), object(4) *(objects are strings)*
- Memory usage: 642.5+ KB

## Column data types

In [14]:
print(df_laptimes.dtypes)

race_id                   int64
gp_year                   int64
gp_name                  object
gp_round                  int64
driver_id                 int64
driver_name              object
rookie_or_experienced    object
lap_number                int64
lap_time                 object
lap_time_ms               int64
dtype: object


## Missing or null values

In [15]:
df_laptimes.isnull().sum() # No nulls present across the dataset!

race_id                  0
gp_year                  0
gp_name                  0
gp_round                 0
driver_id                0
driver_name              0
rookie_or_experienced    0
lap_number               0
lap_time                 0
lap_time_ms              0
dtype: int64

## Check for duplicates

In [16]:
df_laptimes.duplicated().sum() # No duplicates found

np.int64(0)

## Summary statistics

In [17]:
df_laptimes.describe()

,race_id,gp_year,gp_round,driver_id,lap_number,lap_time_ms
count,8222.000000,8222.000000,8222.000000,8222.000000,8222.000000,8.222000e+03
mean,986.430066,2017.356604,11.509000,602.163221,31.275724,9.330352e+04
std,29.875159,1.401271,4.512304,368.615879,18.877443,4.644372e+04
min,930.000000,2015.000000,5.000000,9.000000,1.000000,6.784700e+04
25%,962.000000,2016.000000,8.000000,13.000000,15.000000,7.915075e+04
50%,993.000000,2018.000000,12.000000,822.000000,30.000000,8.756600e+04
75%,1015.000000,2019.000000,14.000000,840.000000,46.000000,1.001305e+05
max,1029.000000,2019.000000,20.000000,847.000000,78.000000,2.118323e+06


## Filtering

In [18]:
# first, save the number of total laps - to compare how many invalid laps have been removed

num_laps_original = len(df_laptimes)
print(num_laps_original)

8222


In [19]:
# Compute the median lap time for each driver’s race session (defined by gp_year, gp_round and driver_id)
df_laptimes['median_lap_time_ms'] = df_laptimes.groupby(['gp_year', 'gp_round', 'driver_id'])['lap_time_ms'].transform('median')

MAX_FILTER_PCT = 1.05 # multiplier for the maximum lap time for filtering purposes

# Compute the max_lap_time = MAX_FILTER_PCT * median lap time - any times landing outside this is excluded. 
df_laptimes['max_lap_time_ms'] = df_laptimes['median_lap_time_ms'] * MAX_FILTER_PCT

# Keep only laps below the max filter - dynamic filter based on 107% rule
df_laptimes = df_laptimes[df_laptimes['lap_time_ms'] <= df_laptimes['max_lap_time_ms']]

# drop both extra median and max columns
#df_laptimes.drop(columns=['median_lap_time_ms', 'max_lap_time_ms'], inplace=True)

# Reset index for df_laptimes
df_laptimes = df_laptimes.reset_index(drop=True)

# Re-check the dataset after filtering
print(df_laptimes.info())

# save the number of laps
num_laps_filtered = len(df_laptimes)
print(num_laps_filtered)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7175 entries, 0 to 7174
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   race_id                7175 non-null   int64  
 1   gp_year                7175 non-null   int64  
 2   gp_name                7175 non-null   object 
 3   gp_round               7175 non-null   int64  
 4   driver_id              7175 non-null   int64  
 5   driver_name            7175 non-null   object 
 6   rookie_or_experienced  7175 non-null   object 
 7   lap_number             7175 non-null   int64  
 8   lap_time               7175 non-null   object 
 9   lap_time_ms            7175 non-null   int64  
 10  median_lap_time_ms     7175 non-null   float64
 11  max_lap_time_ms        7175 non-null   float64
dtypes: float64(2), int64(6), object(4)
memory usage: 672.8+ KB
None
7175


In [20]:
# Check how many observations have been dropped due to invalid lap times
print(f"Total laps before dropping invalid lap times: {num_laps_original}")  # Original number of laps
print(f"Total laps after dropping invalid lap times: {num_laps_filtered}")  # Number of laps after filtering
print(f"Number of laps dropped due to invalid lap times: {num_laps_original - num_laps_filtered}")  # Calculate the number of laps dropped

Total laps before dropping invalid lap times: 8222
Total laps after dropping invalid lap times: 7175
Number of laps dropped due to invalid lap times: 1047


In [21]:
df_laptimes.describe()

,race_id,gp_year,gp_round,driver_id,lap_number,lap_time_ms,median_lap_time_ms,max_lap_time_ms
count,7175.000000,7175.000000,7175.000000,7175.000000,7175.000000,7175.000000,7175.000000,7175.000000
mean,987.105923,2017.389547,11.502578,607.424669,32.392613,88253.750801,88716.185854,93151.995146
std,29.821768,1.399361,4.486508,366.107320,18.499456,13015.149628,13357.836697,14025.728532
min,930.000000,2015.000000,5.000000,9.000000,1.000000,67847.000000,69019.000000,72469.950000
25%,962.000000,2016.000000,8.000000,13.000000,17.000000,77966.500000,78362.000000,82280.100000
50%,993.000000,2018.000000,12.000000,822.000000,31.000000,86113.000000,86261.500000,90574.575000
75%,1015.000000,2019.000000,14.000000,840.000000,46.000000,96225.000000,96894.000000,101738.700000
max,1029.000000,2019.000000,20.000000,847.000000,78.000000,129712.000000,123583.000000,129762.150000


In [ ]:
# Export the validated lap times dataframe to a new CSV file, called "driver-lap-times-validated.csv"
df_laptimes.to_csv('../processed_data/driver-lap-times-validated.csv')

## Validation conclusion
- No null values found
- Column data types are correct
- 1,047 observations with lap times over 105% of a driver's session median were removed. 
- Proceed with feature engineering using the newly validated and exported "driver-lap-times-validated.csv"